### DISCUSSION TITLE
Confused

### DISCUSSION THREAD

**Discussion by DavidPratten** [geldata/gel#2642] [Q&A] [answered]
Hi,

I'm trying to figure out how these terms relate for edgedb.  Can someone point me to an explanation?

- project
- database
- module
- schema
- instance
- dbschema (folder)
- migrations

e.g. do databases live within instances? Is there one schema per instance or per database? How do modules fit in?  Are migrations applied to databases or to instances?  A project and dbschema folder are entirely filesystem based?

Thanks

David

**Comment by jaclarke** [comment, marked as answer]
The defininitions of all these terms are somewhat spread across the docs, so I'll try to summarise how they fit together here:

At the top level are **instances** which are individual instances of EdgeDB server (of a specific version) running in a particular data directory (https://www.edgedb.com/docs/admin/setup#creating-an-edgedb-instance). Each instance can contain multiple **databases**, and also has instance level config (called 'system' in the docs: https://www.edgedb.com/docs/edgeql/admin/configure) and roles (https://www.edgedb.com/docs/edgeql/admin/roles) that apply to all the databases in that particular instance.

Each **database** has it's own **schema** (https://www.edgedb.com/docs/datamodel/overview) and data. Within the schema there are **modules** (https://www.edgedb.com/docs/datamodel/modules) that just act as namespaces to help group related schema types/functions/etc. There are the following builtin modules, that EdgeDB provides for every database:
```
cal
cfg
default
math
schema
std
stdgraphql
sys
```
The `default` module is usually where you would put your own schema types, but you can organise your schema types into whatever custom modules you want. The `default` (and `std`) modules are special though, since you don't have to specify the module when using them (ie. you can just write `SELECT Movie {...}` instead of `SELECT default::Movie {...}`).

There may be some confusion with the term `instance`, since we also use that term for a slightly different concept of **named instances** in the EdgeDB client libraries (https://www.edgedb.com/docs/clients/00_python/instances), the cli tool (the `-I` option: https://www.edgedb.com/docs/cli/edgedb#options) and in projects. **Named instances** are created by the `edgedb server init <name>` command (https://www.edgedb.com/docs/cli/edgedb_server/edgedb_server_init), and refer to a specific *database* (by default named `edgedb`) within a specific *instance* of EdgeDB that the command initialises. The details needed to connect to this database (eg. database name, instance address and port, and username and password) are stored in a credentials file (in a well known `.edgedb` directory on your system) with the name of the instance name. The idea of this is to allow you to just use a simple name to connect to your database in development (https://www.edgedb.com/blog/edgedb-1-0-alpha-5-luhman#simplified-administration-with-named-instances).

**Edit:** Further clarification on named instances: these aren't a special type of instance, the instance itself isn't aware of the name, and it can have multiple databases in it if you want. The name is just defined by the name of the file containing the credentials to connect to the *default* database in an instance. If you create multiple databases in the instance you can override the database specified in this file, eg. `edgedb -I <instance name> -d <custom database>`. In the common case though the instance will have just one database (we recommend each project has it's own instance, so projects can be updated to new versions of EdgeDB independently, and most projects likely only use one database), so in the CLI and client libraries, an instance name is an alias to the credentials to connect to this default database in an instance.

The **dbschema** directory is just a convention; it's the directory in your project directory where we recommend you put the `.esdl` files that define your schema, and it's where the EdgeDB CLI will look by default for your schema when doing migrations. Which brings us to **migrations**: in your `dbschema` directory you have the declarative schema definition for your project (written in `sdl`: https://www.edgedb.com/docs/edgeql/sdl/index), when you update this schema definition, you then run `edgedb create-migration` which diffs the definitions in `dbschema` with the schema of your actual running database, and generates a migration file (containing the steps to get from the current database schema to your new schema using `ddl`: https://www.edgedb.com/docs/edgeql/ddl/index) which is also stored in the `dbschema/migrations` directory. Finally running `edgedb migrate` will look at these migration files and actually apply them to your running database. (More details here: https://www.edgedb.com/docs/cli/edgedb_migration).

Lastly there is **edgedb project** which essentially ties a named instance to your project directory (using a file in the `.edgedb` directory similar to named instances), so when you use the EdgeDB CLI or connect using the client libraries within your project directory, you will no longer need to even provide the instance name. It also has a benefit of defining the version of EdgeDB you are using, so you can clone your project repo and the CLI can automatically create a named instance running the correct EdgeDB version and link it to your cloned project directory, and since your migrations are stored there in `dbschema`, the CLI can then automatically apply them to the new database (https://www.edgedb.com/blog/introducing-edgedb-projects and https://www.edgedb.com/docs/cli/edgedb_project).

**Reply by DavidPratten** [reply]
Exceptionally helpful. 

Thanks

I will raise another issue around specific elements of this which are confusing.

David

**Reply by 1st1** [reply]
FWIW we'll soon add a top-level Glossary section to our docs to clarify all these concepts in succinct form in one easily discoverable place. Admittedly we also need to work on adding more tutorial and guides.
